resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -505,32 +508,37 @@ private[yarn] class YarnAllocator(
 
       if (numExecutorsRunning < targetNumExecutors) {
         if (launchContainers) {
-          launcherPool.execute(new Runnable {
-            override def run(): Unit = {
-              try {
-                new ExecutorRunnable(
-                  Some(container),
-                  conf,
-                  sparkConf,
-                  driverUrl,
-                  executorId,
-                  executorHostname,
-                  executorMemory,
-                  executorCores,
-                  appAttemptId.getApplicationId.toString,
-                  securityMgr,
-                  localResources
-                ).run()
-                updateInternalState()
-              } catch {
-                case NonFatal(e) =>
-                  logError(s"Failed to launch executor $executorId on container $containerId", e)
-                  // Assigned container should be released immediately to avoid unnecessary resource
-                  // occupation.
-                  amClient.releaseAssignedContainer(containerId)
+          try {
+            numExecutorToBeLaunched += 1
+            launcherPool.execute(new Runnable {
+              override def run(): Unit = {
+                try {
+                  new ExecutorRunnable(
+                    Some(container),
+                    conf,
+                    sparkConf,
+                    driverUrl,
+                    executorId,
+                    executorHostname,
+                    executorMemory,
+                    executorCores,
+                    appAttemptId.getApplicationId.toString,
+                    securityMgr,
+                    localResources
+                  ).run()
+                  updateInternalState()
+                } catch {
+                  case NonFatal(e) =>
+                    logError(s"Failed to launch executor $executorId on container $containerId", e)
+                    // Assigned container should be released immediately
+                    // to avoid unnecessary resource occupation.
+                    amClient.releaseAssignedContainer(containerId)
+                }
               }
-            }
-          })
+            })
+          } finally {
+            numExecutorToBeLaunched -= 1
```

Comment:
Also, shouldn't this be done in the actual thread that launching the executor? Otherwise you're decrementing the counter as soon as the task is submitted for execution, but you're not really waiting for the task to execute.